src/Mod/Sketcher/Gui/DrawSketchHandlerBSplineByInterpolation.h

```diff
@@ -0,0 +1,512 @@
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef SKETCHERGUI_DrawSketchHandlerBSplineByInterpolation_H
+#define SKETCHERGUI_DrawSketchHandlerBSplineByInterpolation_H
+
+#include <QInputDialog>
+#include <Inventor/events/SoKeyboardEvent.h>
+
+#include "GeometryCreationMode.h"
+
+
+namespace SketcherGui {
+
+extern GeometryCreationMode geometryCreationMode; // defined in CommandCreateGeo.cpp
+
+class DrawSketchHandlerBSplineByInterpolation: public DrawSketchHandler
+{
+public:
+    explicit DrawSketchHandlerBSplineByInterpolation(int constructionMethod)
+      : Mode(STATUS_SEEK_FIRST_POINT)
+      , MousePressMode(MOUSE_NOT_PRESSED)
+      , ConstrMethod(constructionMethod)
+      , SplineDegree(3)
+      , IsClosed(false)
+    {
+        addSugConstraint();
+        applyCursor();
+    }
+
+    virtual ~DrawSketchHandlerBSplineByInterpolation() = default;
+
+    /// modes
+    enum SELECT_MODE {
+        STATUS_SEEK_FIRST_POINT,
+        STATUS_SEEK_ADDITIONAL_POINTS,
+        STATUS_CLOSE
+    };
+
+    // TODO: this kind of behavior will be useful in a superclass
+    // when LMB is pressed it's a transitional state so some undos can't be done
+    // (like delete last knot)
+    enum MOUSE_PRESS_MODE {
+        MOUSE_PRESSED,
+        MOUSE_NOT_PRESSED
+    };
+
+    void mouseMove(Base::Vector2d onSketchPos) override
+    {
+        prevCursorPosition = onSketchPos;
+
+        if (Mode==STATUS_SEEK_FIRST_POINT) {
+            setPositionText(onSketchPos);
+
+            if (seekAutoConstraint(sugConstr.back(), onSketchPos, Base::Vector2d(0.f,0.f))) {
+                renderSuggestConstraintsCursor(sugConstr.back());
+                return;
+            }
+        }
+        else if (Mode==STATUS_SEEK_ADDITIONAL_POINTS) {
+
+            drawControlPolygonToPosition(onSketchPos);
+
+            drawCursorToPosition(onSketchPos);
+
+            if (seekAutoConstraint(sugConstr.back(), onSketchPos, Base::Vector2d(0.f,0.f))) {
+                renderSuggestConstraintsCursor(sugConstr.back());
+                return;
+            }
+        }
+    }
+
+    bool pressButton(Base::Vector2d onSketchPos) override
+    {
+        prevCursorPosition = onSketchPos;
+
+        MousePressMode = MOUSE_PRESSED;
+
+        if (Mode == STATUS_SEEK_FIRST_POINT) {
+            BSplineKnots.push_back(onSketchPos);
+
+            Mode = STATUS_SEEK_ADDITIONAL_POINTS;
+
+            // insert circle point for knot, defer internal alignment constraining.

```

Comment:
circle?

Reply:
Will take a look. Most of these are discrepancies in the comments so they shouldn't have an effect on the result, but of course should be fixed to avoid confusion.

Reply:
Reworded all mentions of circles and poles.